### PR TITLE
test streaming accuracy python implementation

### PR DESCRIPTION
**Pull request by zhizunbao-y**

### CODE REVIEW

**Comment by zhizunbao-y** [comment]
I signed it!

**Comment by tafsiri** [comment]
@gbaned did you mean to request a review from someone else? I don't generally work on the python tf implementation.

**Comment by chsigg** [comment]
@gbaned, I would like to skip reviewing this CL. I'm not familiar with this code.

**Comment by jdduke** [comment]
Neither am I a good reviewer for this CL.

**Comment by gbaned** [comment]
@yifeif Can you please review this PR ?

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

File: tensorflow/examples/speech_commands/recognize_commands.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+        # Add the latest result to the head of the deque.
+        self._previous_results.append([current_time_ms, latest_results])
+
+        # Prune any earlier results that are too old for the averaging window.
+        time_limit = current_time_ms - self._average_window_duration_ms
+        while time_limit > self._previous_results[0][0]:
+            self._previous_results.popleft()
+
+        # If there are to few results, assume the result will be unreliable and bail.
```
Comment (frankchn) [inline tensorflow/examples/speech_commands/recognize_commands.py]: typo: `too few`

File: tensorflow/examples/speech_commands/recognize_commands.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
+from collections import deque
+import numpy as np
+
+class RecognizeResult:
```
Comment (frankchn) [inline tensorflow/examples/speech_commands/recognize_commands.py]: Here, and other places: Inherit classes from `object`, i.e. `class RecognizeResult(object):`

File: tensorflow/examples/speech_commands/test_streaming_accuracy.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
+import logging
+import soundfile as sf
+import tensorflow as tf
+import numpy as np
+from recognize_commands import RecognizeCommands, RecognizeResult
+from accuracy_utils import StreamingAccuracyStats
+
+
+# Reads a model graph definition from disk, and creates a default graph object
+
```
Comment (frankchn) [inline tensorflow/examples/speech_commands/test_streaming_accuracy.py]: Remove extra line

File: tensorflow/examples/speech_commands/test_streaming_accuracy.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        with tf.gfile.GFile(mode_file, 'rb') as fid:
+            serialized_graph = fid.read()
+            od_graph_def.ParseFromString(serialized_graph)
+            tf.import_graph_def(od_graph_def, name='')
+    return graph
+
+
+# Takes a file name, and loads a list of labels from it, one per line, and
+# returns a vector of strings.
+
```
Comment (frankchn) [inline tensorflow/examples/speech_commands/test_streaming_accuracy.py]: Remove extra line

File: tensorflow/examples/speech_commands/test_streaming_accuracy.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+def read_label_file(file_name):
+    label_list = []
+    with open(file_name, 'r') as f:
+        lines = f.readlines()
+        for line in lines:
+            label_list.append(line.strip())
+    return label_list
+
+# Takes a file name, loads a piece of wav from it, and return sample_rate and numpy data of np.float64
+
```
Comment (frankchn) [inline tensorflow/examples/speech_commands/test_streaming_accuracy.py]: Remove extra line

File: tensorflow/examples/speech_commands/test_streaming_accuracy.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+def read_wav_file(file_name):
+    data, sample_rate = sf.read(file_name,dtype='float32')
+    return sample_rate, data
+
+
+def main():
+    logging.basicConfig(level=logging.INFO)
+
+    wav_file = "/home/yuan/datasets/speech_dataset/generated_wav_and_label/streaming_test.wav"
```
Comment (frankchn) [inline tensorflow/examples/speech_commands/test_streaming_accuracy.py]: Can you take this in as a parameter (using absl flags perhaps). Same with everything else

File: tensorflow/examples/speech_commands/test_streaming_accuracy.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    ground_truth_file = '/home/yuan/datasets/speech_dataset/generated_wav_and_label/streaming_test_labels.txt'
+
+    label_file = "/home/yuan/speech_command_result/speech_commands_train/conv_labels.txt"
+    model_file = "/home/yuan/speech_command_result/model/frozen_graph.pb"
+
+    #  ops = tf.get_default_graph().get_operations()
```
Comment (frankchn) [inline tensorflow/examples/speech_commands/test_streaming_accuracy.py]: Remove debugging

File: tensorflow/examples/speech_commands/test_streaming_accuracy.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    #  all_tensor_names = {output.name for op in ops for output in op.outputs}
+    input_data_name = "decoded_sample_data:0"
+    input_rate_name = "decoded_sample_data:1"
+    output_softmax_label_name = 'labels_softmax:0'
+
+    clip_duration_ms = 1000
+    clip_stride_ms = 30
+
+    average_window_duration_ms = 500
```
Comment (frankchn) [inline tensorflow/examples/speech_commands/test_streaming_accuracy.py]: You might also want this as parameters.

**Comment by frankchn** [comment]
@zhizunbao-y 

Thanks for the PR! In addition to my comments in the doc, can you also do the following:

1. Add your script to the BUILD file so that folks can run this as part of the bazel build environment
2. Add instructions to your main file (test_stream_accuracy.py) detailing how to run this

**Review by frankchn — CHANGES_REQUESTED** [review]

**Comment by gbaned** [comment]
@zhizunbao-y Did you get a chance to look on reviewer comments? Please let us know on the update. Thanks!

**Comment by zhizunbao-y** [comment]
> @zhizunbao-y Did you get a chance to look on reviewer comments? Please let us know on the update. Thanks!
Okay, thank you for informing me. I will revise and update it as soon as possible.

**Comment by zhizunbao-y** [comment]
I have done the refinement followed by @frankchn .

File: tensorflow/examples/speech_commands/accuracy_utils.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+    else:
+      any_match_percentage = self._how_many_gt_matched / self._how_many_gt * 100
+      correct_match_percentage = self._how_many_c / self._how_many_gt * 100
+      wrong_match_percentage = self._how_many_w / self._how_many_gt * 100
+      false_positive_percentage = self._how_many_fp / self._how_many_gt * 100
+      tf.logging.info('{:.1f}% matched, {:.1f}% correct, {:.1f}% wrong, '
+                   '{:.1f}% false positive'.format(any_match_percentage,
+                                                   correct_match_percentage,
+                                                   wrong_match_percentage,
+                                                   false_positive_percentage))
```
Comment (frankchn) [inline tensorflow/examples/speech_commands/accuracy_utils.py]: Add trailing new line to the file.

File: tensorflow/examples/speech_commands/test_streaming_accuracy.py (hunk 4 above)
Comment (frankchn) [reply tensorflow/examples/speech_commands/test_streaming_accuracy.py]: Extra line still there

**Review by frankchn — DISMISSED** [review]
LGTM, except 2 minor nits. Thanks for doing this :)

**Comment by gbaned** [comment]
@zhizunbao-y Could you please check reviewer comments? Thanks!

**Review by frankchn — APPROVED** [review]

**Comment by gbaned** [comment]
@zhizunbao-y Could you please address Ubuntu Sanity errors? Thanks!

**Review by frankchn — DISMISSED** [review]

**Comment by gbaned** [comment]
@zhizunbao-y Still, Ubuntu Sanity checks failing. Could you please check again?  Thanks!

**Comment by zhizunbao-y** [comment]
@gbaned I don't know which part of the code made the Ubuntu sanity check failing. Please give me some guidance.

**Comment by zhizunbao-y** [comment]
I am sorry. It might be an input type error that caused this check error. I have fixed it. Please run check again. Thanks!

**Comment by frankchn** [comment]
Here are the errors from Ubuntu Sanity:

From pylint:
```
tensorflow/examples/speech_commands/recognize_commands.py:107: [C0330(bad-continuation), ] Wrong continued indentation (remove 1 space).
tensorflow/examples/speech_commands/recognize_commands.py:128: [C0330(bad-continuation), ] Wrong hanging indentation (remove 35 spaces).
tensorflow/examples/speech_commands/recognize_commands.py:129: [C0330(bad-continuation), ] Wrong hanging indentation (remove 28 spaces).
tensorflow/examples/speech_commands/recognize_commands.py:135: [C0330(bad-continuation), ] Wrong continued indentation (add 10 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:136: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:138: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:140: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:148: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:149: [C0330(bad-continuation), ] Wrong continued indentation (add 9 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:154: [C0330(bad-continuation), ] Wrong continued indentation (remove 1 space).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:169: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:172: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:173: [C0330(bad-continuation), ] Wrong continued indentation (remove 5 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:174: [C0330(bad-continuation), ] Wrong continued indentation (remove 5 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:175: [C0330(bad-continuation), ] Wrong continued indentation (remove 5 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:184: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:185: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:186: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:187: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:189: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:190: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:191: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:192: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:194: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:195: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:196: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:197: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:199: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:200: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:201: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:202: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:204: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:205: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:206: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:207: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:207: [C0326(bad-whitespace), ] Exactly one space required after comma
tensorflow/examples/speech_commands/test_streaming_accuracy.py:208: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:210: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:211: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:212: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:213: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:215: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:216: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:217: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:218: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:220: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:221: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:222: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:223: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:225: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:226: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:227: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:228: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:230: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:231: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:232: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:233: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:235: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:236: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:237: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:238: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:240: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:241: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:242: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:243: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:246: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:247: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:248: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/test_streaming_accuracy.py:249: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/accuracy_utils.py:64: [C0330(bad-continuation), ] Wrong continued indentation (remove 10 spaces).
tensorflow/examples/speech_commands/accuracy_utils.py:144: [C0330(bad-continuation), ] Wrong hanging indentation (add 2 spaces).
tensorflow/examples/speech_commands/accuracy_utils.py:151: [C0330(bad-continuation), ] Wrong continued indentation (add 3 spaces).
```

From Ubuntu Sanity:
```
FAIL: buildifier found errors and/or warnings in above BUILD files.
buildifier suggested the following changes:
tensorflow/examples/speech_commands/BUILD:
16,17c16,17
<     main="accuracy_utils.py",
<     srcs = ["accuracy_utils.py",],
---
>     srcs = ["accuracy_utils.py"],
>     main = "accuracy_utils.py",
23c23
<         ],
---
>     ],
28,29c28,29
<     main="recognize_commands.py",
<     srcs = ["recognize_commands.py",],
---
>     srcs = ["recognize_commands.py"],
>     main = "recognize_commands.py",
35c35
<         ],
---
>     ],
39a40
>     srcs = ["test_streaming_accuracy.py"],
41d41
<     srcs = ["test_streaming_accuracy.py",],
50c50
<         ],
---
>     ],
Please fix manually or run buildifier <file> to auto-fix.
```

**Comment by zhizunbao-y** [comment]
I used the pylintrc in tensorflow/tensorflow/tools/ci_build/ to check my code.
This time might be ok.

**Review by frankchn — APPROVED** [review]

**Pull request merged**